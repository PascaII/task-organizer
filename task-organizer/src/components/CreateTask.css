.create-task {
    max-width: 1000px;
    margin: 0 auto;
    padding: 20px;
}
.FormControl{
    display: flex;
    justify-content: space-between;
    max-width: 1200px;
}
.FormGroup{
    margin-bottom: 1em;
}
.form-control{
    max-width: 90%;
}
.table{
    max-width: 1200px;
}
td{
    text-align: left;
}
.ButtonMargin{
    margin-bottom: 2em;
}

.CardMargin, .FormControl{
    margin: auto;
}

.CardMargin{
    width: 18rem;
}