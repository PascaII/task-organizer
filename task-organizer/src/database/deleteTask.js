import { getFirestore, doc, deleteDoc } from "firebase/firestore";
import config from "./firebase-config";
import { initializeApp } from "firebase/app";
import getConfig from "../configuration/config-helper";
import logger from "../components/logger";

const firebaseConfig = {
    apiKey: config.apiKey,
    authDomain: config.authDomain,
    projectId: config.projectId,
    storageBucket: config.storageBucket,
    messagingSenderId: config.messagingSenderId,
    appId: config.appId,
    measurementId: config.measurementId,
};

const app = initializeApp(firebaseConfig);
const db = getFirestore(app);

// function that deletes entry from firestore based on its UUID
async function DeleteTaskByID(taskID) {
    try {
        const taskDocRef = doc(db, getConfig().collection, taskID);
        await deleteDoc(taskDocRef);
        logger.info(`Task with ID ${taskID} deleted successfully.`);
    } catch (error) {
        logger.error(`Error deleting task: ${error}`);
    }
}

export { DeleteTaskByID };
