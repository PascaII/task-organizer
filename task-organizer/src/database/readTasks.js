import { getFirestore, collection, getDocs, doc, getDoc } from "firebase/firestore";
import config from "./firebase-config";
import { initializeApp } from "firebase/app";
import getConfig from "../configuration/config-helper";

const firebaseConfig = {
    apiKey: config.apiKey,
    authDomain: config.authDomain,
    projectId: config.projectId,
    storageBucket: config.storageBucket,
    messagingSenderId: config.messagingSenderId,
    appId: config.appId,
    measurementId: config.measurementId,
};

const app = initializeApp(firebaseConfig);
const db = getFirestore(app);

// function that returns array of all tasks
async function ReadAllTasks() {
    const tasksCollectionRef = collection(db, getConfig().collection);
    const tasksSnapshot = await getDocs(tasksCollectionRef);

    const tasks = [];
    tasksSnapshot.forEach((doc) => {
        tasks.push({ id: doc.id, ...doc.data() });
    });
    console.log(tasks);
    return tasks;
}

// function that returns single task by its UUID
async function ReadTaskByID(taskId) {
    const taskDocRef = doc(db, getConfig().collection, taskId);
    const taskDocSnapshot = await getDoc(taskDocRef);

    if (taskDocSnapshot.exists()) {
        const taskData = { id: taskDocSnapshot.id, ...taskDocSnapshot.data() };
        console.log(taskData);
        return taskData;
    } else {
        console.log("Task not found");
        return null;
    }
}

export { ReadAllTasks, ReadTaskByID };
