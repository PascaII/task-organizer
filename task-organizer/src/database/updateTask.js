import { getFirestore, doc, updateDoc } from "firebase/firestore";
import config from "./firebase-config";
import { initializeApp } from "firebase/app";
import getConfig from "../configuration/config-helper";

const firebaseConfig = {
    apiKey: config.apiKey,
    authDomain: config.authDomain,
    projectId: config.projectId,
    storageBucket: config.storageBucket,
    messagingSenderId: config.messagingSenderId,
    appId: config.appId,
    measurementId: config.measurementId,
};

const app = initializeApp(firebaseConfig);
const db = getFirestore(app);

// function to update a task in firestore using its UUID
async function UpdateTaskByID(taskID, name, description, status, priority) {
    const taskRef = doc(db, getConfig().collection, taskID);

    try {
        await updateDoc(taskRef, {
            name: name,
            description: description,
            status: status,
            priority: priority,
        });
        console.log("Task updated successfully!");
    } catch (error) {
        console.error("Error updating task:", error);
    }
}

export { UpdateTaskByID };
