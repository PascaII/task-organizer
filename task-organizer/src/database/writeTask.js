import { doc, getFirestore, setDoc } from "firebase/firestore";
import config from "./firebase-config";
import { initializeApp } from "firebase/app";
import getConfig from "../configuration/config-helper";
import { v4 as uuidv4 } from 'uuid'; // Import the uuid library

const firebaseConfig = {
    apiKey: config.apiKey,
    authDomain: config.authDomain,
    projectId: config.projectId,
    storageBucket: config.storageBucket,
    messagingSenderId: config.messagingSenderId,
    appId: config.appId,
    measurementId: config.measurementId,
};

const app = initializeApp(firebaseConfig);
const db = getFirestore(app);

// function that writes a task into the firestore and returns its UUID
function WriteTask(name, description, status, priority) {
    const id = uuidv4(); // Generate a new UUID

    setDoc(doc(db, getConfig().collection, id), {
        id: id,
        name: name,
        description: description,
        status: status,
        priority: priority,
    });

    return id;
}

export default WriteTask;
