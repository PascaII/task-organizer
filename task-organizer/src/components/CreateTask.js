import React, { useState } from "react";
import "./CreateTask.css";
import WriteTask from "../database/writeTask";
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Table from 'react-bootstrap/Table';
import Card from 'react-bootstrap/Card';
import logger from "./logger";

function CreateTask() {
    const [description, setDescription] = useState("");
    const [name, setName] = useState("");
    const [priority, setPriority] = useState("");
    const [status, setStatus] = useState("");
    const priorityOptions = ["High", "Medium", "Low"];
    const statusOptions = ["To do", "In progress", "Done"];
    const [submissionStatus, setSubmissionStatus] = useState(null);

    // function to handle the selected status
    const handleStatusSelect = (option) => {
        setStatus(option);
    };

    // function to handle the selected priority
    const handlePrioritySelect = (option) => {
        setPriority(option)
    }

    // function to submit form
    const handleSubmit = (e) => {
        e.preventDefault();
        try {
            let docId = WriteTask(name, description, status, priority);
            setSubmissionStatus("success");
            setDescription("");
            setName("");
            setPriority("");
            setStatus("");
            logger.info(`Task created with the following properties: ${docId}, ${description}, ${priority}, ${name}, ${status}`);
        } catch (error) {
            logger.error(`Error creating task: ${error}`);
            setSubmissionStatus("error");
        }
    };


    return (
        <div className="create-task">
            <div className="container">
                <h1>Create a task</h1>
                <br />
                <div className="form">
                    <Form onSubmit={handleSubmit}>
                        <Table bordered={false} striped={false}>
                            <tbody>
                                <tr>
                                    <td>
                                        <Form.Group>
                                            <Form.Label htmlFor="name">Name:</Form.Label>
                                        </Form.Group>
                                    </td>
                                    <td>
                                        <Form.Control
                                            type="text"
                                            id="name"
                                            name="name"
                                            value={name}
                                            onChange={(e) => setName(e.target.value)}
                                            required
                                        />
                                    </td>
                                </tr>
                                <tr>
                                    <td>
                                        <Form.Group>
                                            <Form.Label htmlFor="description">Description:</Form.Label>
                                        </Form.Group>
                                    </td>
                                    <td>
                                        <Form.Control
                                            id="description"
                                            name="description"
                                            type="text"
                                            value={description}
                                            onChange={(e) => setDescription(e.target.value)}
                                            required
                                        />
                                    </td>
                                </tr>
                                <tr>
                                    <td>
                                        <Form.Group>
                                            <Form.Label htmlFor="priority">Priority:</Form.Label>
                                        </Form.Group>
                                    </td>
                                    <td>
                                        <Form.Select
                                            id="priority"
                                            name="priority"
                                            value={priority}
                                            onChange={(e) => handlePrioritySelect(e.target.value)}
                                            required
                                        >
                                            <option value="" disabled>
                                                Select an option
                                            </option>
                                            {priorityOptions.map((option) => (
                                                <option key={option} value={option}>
                                                    {option}
                                                </option>
                                            ))}
                                        </Form.Select>
                                    </td>
                                </tr>
                                <tr>
                                    <td>
                                        <Form.Group>
                                            <Form.Label htmlFor="status">Status:</Form.Label>
                                        </Form.Group>
                                    </td>
                                    <td>
                                        <Form.Select
                                            id="status"
                                            name="status"
                                            value={status}
                                            onChange={(e) => handleStatusSelect(e.target.value)}
                                            required
                                        >
                                            <option value="" disabled>
                                                Select an option
                                            </option>
                                            {statusOptions.map((option) => (
                                                <option key={option} value={option}>
                                                    {option}
                                                </option>
                                            ))}
                                        </Form.Select>
                                    </td>
                                </tr>
                            </tbody>
                        </Table>
                        <Button className="ButtonMargin" variant="primary" type="submit" style={{ marginBottom: "2em" }}>
                            Submit
                        </Button>
                        {submissionStatus === "success" && (
                            <Card className="CardStyle">
                                <Card.Body>
                                    <Card.Title className="text-success">Success</Card.Title>
                                    <Card.Text>
                                        Task successfully created!
                                    </Card.Text>
                                </Card.Body>
                            </Card>
                        )}
                        {submissionStatus === "error" && (
                            <Card className="CardStyle">
                                <Card.Body>
                                    <Card.Title className="text-danger">Error</Card.Title>
                                    <Card.Text>
                                        Error while creating task. Please try again!
                                    </Card.Text>
                                </Card.Body>
                            </Card>
                        )}
                    </Form>
                </div>
            </div>
        </div>

    );
}

export default CreateTask;
