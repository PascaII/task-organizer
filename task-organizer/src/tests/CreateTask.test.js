import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import CreateTask from '../components/CreateTask';
import renderer from "react-test-renderer";

describe('CreateTask Component', () => {
  test('renders CreateTask component', () => {
    render(<CreateTask />);

    // Check if the heading is rendered
    expect(screen.getByText('Create a task')).toBeInTheDocument();

    // Check if form elements are rendered
    expect(screen.getByLabelText('Name:')).toBeInTheDocument();
    expect(screen.getByLabelText('Description:')).toBeInTheDocument();
    expect(screen.getByLabelText('Priority:')).toBeInTheDocument();
    expect(screen.getByLabelText('Status:')).toBeInTheDocument();

    // Check if the submit button is rendered
    expect(screen.getByText('Submit')).toBeInTheDocument();
  });

  test('handles form submission', () => {
    render(<CreateTask />);

    // Mocking the WriteTask function
    const originalWriteTask = require('../database/writeTask').default;
    const mockWriteTask = jest.fn();
    require('../database/writeTask').default = mockWriteTask;

    // Fill out the form
    fireEvent.change(screen.getByLabelText('Name:'), { target: { value: 'Task Name' } });
    fireEvent.change(screen.getByLabelText('Description:'), { target: { value: 'Task Description' } });
    fireEvent.change(screen.getByLabelText('Priority:'), { target: { value: 'High' } });
    fireEvent.change(screen.getByLabelText('Status:'), { target: { value: 'To do' } });

    // Submit the form
    fireEvent.click(screen.getByText('Submit'));

    // Check if WriteTask is called with the correct arguments
    expect(mockWriteTask).toHaveBeenCalledWith('Task Name', 'Task Description', 'To do', 'High');

    // Restore the original WriteTask function
    require('../database/writeTask').default = originalWriteTask;
  });
  test('snapshot matches implementation', () => {
    const tree = renderer.create(<CreateTask />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
