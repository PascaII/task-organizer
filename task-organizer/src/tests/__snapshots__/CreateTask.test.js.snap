// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTask Component snapshot matches implementation 1`] = `
<div
  className="create-task"
>
  <div
    className="container"
  >
    <h1>
      Create a task
    </h1>
    <br />
    <div
      className="form"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <table
          className="table"
        >
          <tbody>
            <tr>
              <td>
                <div>
                  <label
                    className="form-label"
                    htmlFor="name"
                  >
                    Name:
                  </label>
                </div>
              </td>
              <td>
                <input
                  className="form-control"
                  id="name"
                  name="name"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value=""
                />
              </td>
            </tr>
            <tr>
              <td>
                <div>
                  <label
                    className="form-label"
                    htmlFor="description"
                  >
                    Description:
                  </label>
                </div>
              </td>
              <td>
                <input
                  className="form-control"
                  id="description"
                  name="description"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value=""
                />
              </td>
            </tr>
            <tr>
              <td>
                <div>
                  <label
                    className="form-label"
                    htmlFor="priority"
                  >
                    Priority:
                  </label>
                </div>
              </td>
              <td>
                <select
                  className="form-select"
                  id="priority"
                  name="priority"
                  onChange={[Function]}
                  required={true}
                  value=""
                >
                  <option
                    disabled={true}
                    value=""
                  >
                    Select an option
                  </option>
                  <option
                    value="High"
                  >
                    High
                  </option>
                  <option
                    value="Medium"
                  >
                    Medium
                  </option>
                  <option
                    value="Low"
                  >
                    Low
                  </option>
                </select>
              </td>
            </tr>
            <tr>
              <td>
                <div>
                  <label
                    className="form-label"
                    htmlFor="status"
                  >
                    Status:
                  </label>
                </div>
              </td>
              <td>
                <select
                  className="form-select"
                  id="status"
                  name="status"
                  onChange={[Function]}
                  required={true}
                  value=""
                >
                  <option
                    disabled={true}
                    value=""
                  >
                    Select an option
                  </option>
                  <option
                    value="To do"
                  >
                    To do
                  </option>
                  <option
                    value="In progress"
                  >
                    In progress
                  </option>
                  <option
                    value="Done"
                  >
                    Done
                  </option>
                </select>
              </td>
            </tr>
          </tbody>
        </table>
        <button
          className="ButtonMargin btn btn-primary"
          disabled={false}
          style={
            Object {
              "marginBottom": "2em",
            }
          }
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
